src/py_string_cache.rs

```diff
@@ -0,0 +1,36 @@
+use pyo3::once_cell::GILOnceCell;
+use pyo3::prelude::*;
+use pyo3::types::PyString;
+use std::sync::{Arc, Mutex};
+
+use ahash::AHashMap;
+
+type PyStringCache = Arc<Mutex<AHashMap<String, Py<PyString>>>>;
+static PY_STRING_CACHE: GILOnceCell<PyStringCache> = GILOnceCell::new();
```

Comment:
@davidhewitt I wonder if we can add a variant of the `get_mut` method to `GILOnceCell` to eliminate the `Mutex`?

```rust
pub fn get_mut_with_gil(&mut self, _py: Python) -> Option<&mut T> {
    // Safe because GIL is held, so no other thread can be writing to this cell concurrently.
}
```

Reply:
That would be useful.

Do you know of anyway to hack a solution to this at present?

Reply:
[`get_mut()`](https://docs.rs/pyo3/latest/pyo3/once_cell/struct.GILOnceCell.html#method.get_mut) requires a mutable reference to `self` while `get()` only gives out immutable reference, so no unless you use a concurrent hashmap, but I don't think concurrent hashmap is useful here since we have GIL.